src/haywire/http_response.c

```diff
@@ -66,13 +78,30 @@ hw_string* create_response_buffer(hw_http_response* response)
 
     int i = 0;
 
-    response_string->value = calloc(1024, 1);
+    char length_header[] = "Content-Length: ";
+    int line_sep_size = sizeof(CRLF);
+
+    int header_buffer_incr = 512;
+    int body_size = resp->body.length + line_sep_size;
+    int header_size_remaining = header_buffer_incr;
+    int response_size = header_size_remaining + sizeof(length_header) + num_chars(resp->body.length) + 2 * line_sep_size + body_size + line_sep_size;
+
+    response_string->value = malloc(response_size);
+
     response_string->length = 0;
     append_string(response_string, cached_entry);
     
     for (i=0; i< resp->number_of_headers; i++)
     {
         http_header header = resp->headers[i];
+
+        header_size_remaining -= strlen(header.name.value) + 2 + strlen(header.value.value) + line_sep_size;
```

Comment:
Change the `strlen(header.name.value)` to `header.name.length` and change `strlen(header.value.value)` to `header.value.length`. Haywire has `hw_string` that has pre-set lengths so that we can avoid string loops inside Haywire. This was a big gain in performance as we avoid looping buffers as much as possible :)

Making this change gave me the following.

```
../bin/wrk/wrk --script ./pipelined_get.lua --latency -d 10s -t 40 -c 760 http://server:8000 -- 32
Running 10s test @ http://server:8000
  40 threads and 760 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.92ms    4.10ms 206.24ms   91.91%
    Req/Sec   234.34k    38.82k  445.76k    85.05%
  Latency Distribution
     50%    1.24ms
     75%    1.95ms
     90%    4.10ms
     99%    0.00us
  93564882 requests in 10.10s, 13.76GB read
Requests/sec: 9264711.05
Transfer/sec:      1.36GB
```

Next we should test unaligned vs aligned buffer sizes but this is looking very promising. We got dynamic buffer sizes without losing noticeable performance :)